tests/plugin_filter/util.py

```diff
@@ -0,0 +1,218 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from qhana_plugin_registry.db.models.plugins import RAMP, PluginTag
+from qhana_plugin_registry.db.models.templates import UiTemplate, TemplateTab
+from qhana_plugin_registry.tasks.plugin_filter import apply_filter_for_tab
+
+import random
+from hypothesis import strategies as st
+from packaging.specifiers import Specifier
+import json
+from sqlalchemy.sql import exists
+
+
+def filter_strategy():
+    return st.one_of(
+        st.fixed_dictionaries({"tag": st.text()}),
+        st.fixed_dictionaries({"name": st.text()}),
+        st.fixed_dictionaries(
+            {"version": st.from_regex(Specifier._regex)}
+        ),  # TODO: type PEP 0440 version specifier
+        st.fixed_dictionaries({"and": st.lists(st.deferred(filter_strategy))}),
+        st.fixed_dictionaries({"or": st.lists(st.deferred(filter_strategy))}),
+        st.fixed_dictionaries({"not": st.deferred(filter_strategy)}),
+    )
+
+
+def extract_key(key: str, filter_dict: dict):
+    if key in filter_dict:
+        yield filter_dict[key]
+    for k, v in filter_dict.items():
+        if k in ["and", "or"]:
+            for sub_filter in v:
+                extract_key(key, sub_filter)
+        if k == "not":
+            extract_key(key, v)
+
+
+def create_template(
+    tmp_db, name: str = "test_template", description: str = "test template"
+) -> str:
+    template_exists = tmp_db.session.query(
+        exists().where(UiTemplate.name == name)
+    ).scalar()
+    if template_exists:
+        return tmp_db.session.query(UiTemplate.id).first()[0]
+    template = UiTemplate(
+        name=name,
+        description=description,
+    )
+    tmp_db.session.add(template)
+    tmp_db.session.commit()
+    template_id = (
+        tmp_db.session.query(UiTemplate.id)
+        .filter(UiTemplate.name == template.name)
+        .first()[0]
+    )
+    return template_id
+
+
+def create_or_update_plugin(
+    tmp_db,
+    name: str = "test-plugin",
+    version: str = "0.0.0",
+    tags: list[PluginTag] | None = None,
+    description: str = "descr",
+) -> RAMP:
+    if not tags:
+        tags = []
+    plugin_exists = tmp_db.session.query(
+        exists().where(RAMP.name == name, RAMP.version == version)
+    ).scalar()
+    if plugin_exists:
+        plugin = (
+            tmp_db.session.query(RAMP)
+            .filter(RAMP.name == name, RAMP.version == version)
+            .first()
+        )
+        plugin.plugin_id = f"{name}@{version}"
```

Comment:
name and version of an existing plugin should never change in the database. In fact, your filters ensure that name and version are already the values you assign here. Just remove the lines that cannot have an effect